crates/sui-transactional-test-runner/src/test_adapter.rs

```diff
@@ -595,11 +604,151 @@ impl<'a> MoveTestAdapter<'a> for SuiTestAdapter<'a> {
+                let data = data.ok_or_else(|| {
+                    anyhow::anyhow!(
+                        "Expected a module text block following 'upgrade' starting on lines {}-{}",
+                        start_line,
+                        command_lines_stop
+                    )
+                })?;
+
+                let state = self.compiled_state();
+                let (modules, warnings_opt) = match syntax {
+                    SyntaxChoice::Source => {
+                        let (units, warnings_opt) = compile_source_units(state, data.path())?;
+                        let modules = units
+                            .into_iter()
+                            .map(|unit| match unit {
+                                AnnotatedCompiledUnit::Module(annot_module) => {
+                                    let (named_addr_opt, _id) = annot_module.module_id();
+                                    let named_addr_opt = named_addr_opt.map(|n| n.value);
+                                    let module = annot_module.named_module.module;
+                                    (named_addr_opt, module)
+                                }
+                                AnnotatedCompiledUnit::Script(_) => panic!(
+                                    "Expected a module text block, not a script, \
+                                    following 'upgrade' starting on lines {}-{}",
+                                    start_line, command_lines_stop
+                                ),
+                            })
+                            .collect();
+                        (modules, warnings_opt)
+                    }
+                    SyntaxChoice::IR => {
+                        let module = compile_ir_module(state, data.path())?;
+                        (vec![(None, module)], None)
+                    }
+                };
+                let (output, mut modules) =
+                    self.upgrade_package(package, modules, upgrade_capability, gas_budget)?;
+                match syntax {
+                    SyntaxChoice::Source => {
+                        let path = data.path().to_str().unwrap().to_owned();
+                        self.compiled_state()
+                            .add_with_source_file(modules, (path, data))
+                    }
+                    SyntaxChoice::IR => {
+                        let module = modules.pop().unwrap().1;
+                        self.compiled_state()
+                            .add_and_generate_interface_file(module);
+                    }
+                };
+                Ok(merge_output(warnings_opt, output))
+            }
         }
     }
 }
 
+fn merge_output(left: Option<String>, right: Option<String>) -> Option<String> {
+    match (left, right) {
+        (None, right) => right,
+        (left, None) => left,
+        (Some(mut left), Some(right)) => {
+            left.push_str(&right);
+            Some(left)
+        }
+    }
+}
+
+macro_rules! move_call {
```

Comment:
Let's move this to a more central place seeing as it's used across the `move_package_upgrade_tests` as well.